### PR TITLE
Feature/12537 fss tooling add download button to search page

### PR DESCRIPTION
**Pull request by Rashi14748**
PR contains implementation and unit tests for file download

### CODE REVIEW

File: FileShareClient/FileShareApiClient.cs [hunk 1]
```diff
@@ -106,6 +112,50 @@ public async Task<Stream> DownloadFileAsync(string batchId, string filename)
...
+                using (var httpRequestMessage = new HttpRequestMessage(HttpMethod.Get, uri))
+                {
+                    var httpClient = await GetAuthenticationHeaderSetClient();
+                    if (fileSizeInBytes != 0 && rangeHeader != null)
+                    {
+                        httpRequestMessage.Headers.Add("Range", rangeHeader);
+                        httpStatusCode = HttpStatusCode.PartialContent;
+                    }
+
+                    var response = await httpClient.SendAsync(httpRequestMessage, CancellationToken.None);
```
Comment (crdandoukho) [inline FileShareClient/FileShareApiClient.cs:137]: Please use the `cancellationToken` parameter here instead of `CancellationToken.None`

Comment (Rashi14748) [reply FileShareClient/FileShareApiClient.cs]: update the same. Thanks

**Review by crdandoukho — CHANGES_REQUESTED** [review]
Please add a test to cover DownloadFileAsync

File: Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs [hunk 2]
```diff
@@ -126,5 +127,69 @@ public async Task TestDownloadFileSetsAuthorizationHeader()
...
+        [Test]
+        public async Task TestBasicDownloadFileWithCancellationToken()
+        {
+            var batchId = Guid.NewGuid().ToString();
+            var downldLocation = @"D";
+            var expectedBytes = Encoding.UTF8.GetBytes("Contents of a file.");
+            nextResponse = new MemoryStream(expectedBytes);
+            var result = await fileShareApiClient.DownloadFileAsync(batchId, "AFilename.txt", downldLocation, 10, CancellationToken.None);
+            Assert.IsTrue(result.IsSuccess);
+            Assert.AreEqual($"/basePath/batch/{batchId}/files/AFilename.txt", lastRequestUri.AbsolutePath);
```
Comment (crdandoukho) [inline Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs:141]: Can you please verify that 'D' contains the expected bytes?

**Review by crdandoukho — CHANGES_REQUESTED** [review]
Please add a test to cover the scenario where `fileSizeInBytes` is larger than `maxDownloadBytes`

File: FileShareClient/FileShareApiClient.cs [hunk 3]
```diff
@@ -17,13 +19,17 @@ public interface IFileShareApiClient
         Task<BatchStatusResponse> GetBatchStatusAsync(string batchId);
         Task<BatchSearchResponse> Search(string searchQuery, int? pageSize = null, int? start = null);
         Task<Stream> DownloadFileAsync(string batchId, string filename);
+        Task<IResult<DownloadFileResponse>> DownloadFileAsync(string batchId, string fileName, string downldLocation, long fileSizeInBytes = 0, CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (Rashi14748) [reply FileShareClient/FileShareApiClient.cs]: replaced parameter dwnldlocation with filestream and updated implementation accordingly. Thanks

File: FileShareClient/FileShareApiClient.cs [hunk 4]
```diff
@@ -143,5 +192,16 @@ public async Task<IEnumerable<string>> GetUserAttributesAsync()
 
             return sb.ToString();
         }
+
+        private async Task ProcessContentStream(Stream contentStream, string fileDownloadLocation)
+        {
+            using (var fileStream =  new FileStream(fileDownloadLocation, FileMode.Create, FileAccess.Write, FileShare.ReadWrite))
```
[the remark this answers is not in the document]
Comment (Rashi14748) [reply FileShareClient/FileShareApiClient.cs]: agree, hence removed this code block. Thanks

File: Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs (hunk 2 above)
Comment (Rashi14748) [reply Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs:141]: added test scenario when fileSizeInBytes is larger than maxDownloadBytes.

Comment (Rashi14748) [reply Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs:144]: added test for the downloaded file contains the expected bytes. Thanks for review.

File: FileShareClient/FileShareApiClient.cs [hunk 5]
```diff
@@ -106,6 +112,51 @@ public async Task<Stream> DownloadFileAsync(string batchId, string filename)
             }
         }
 
+        public async Task<IResult<DownloadFileResponse>> DownloadFileAsync(string batchId, string fileName, FileStream destinationStream, long fileSizeInBytes = 0, CancellationToken cancellationToken = default)
```
Comment (crdandoukho) [inline FileShareClient/FileShareApiClient.cs:115]: destinationStream should be Stream not FileStream. There is no reason to restrict the destination to a file. The user of the library might want to stream directly to a MemoryStream, a compressed stream, a network socket, or a zip archive for example.
```suggestion
        public async Task<IResult<DownloadFileResponse>> DownloadFileAsync(string batchId, string fileName, Stream destinationStream, long fileSizeInBytes = 0, CancellationToken cancellationToken = default)
```

File: FileShareClient/FileShareApiClient.cs [hunk 6]
```diff
@@ -106,6 +112,51 @@ public async Task<Stream> DownloadFileAsync(string batchId, string filename)
...
+                startByte = endByte + 1;
+                endByte += maxDownloadBytes;
+
+                if (endByte > fileSizeInBytes - 1)
+                {
+                    endByte = fileSizeInBytes - 1;
+                }
+
+            }
+            destinationStream.Close();
```
Comment (crdandoukho) [inline FileShareClient/FileShareApiClient.cs:156]: Please do not close the stream. The stream is owned by the caller

File: Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs [hunk 7]
```diff
@@ -126,5 +127,116 @@ public async Task TestDownloadFileSetsAuthorizationHeader()
             Assert.AreEqual("bearer", fakeHttpClientFactory.HttpClient.DefaultRequestHeaders.Authorization.Scheme);
             Assert.AreEqual(DUMMY_ACCESS_TOKEN, fakeHttpClientFactory.HttpClient.DefaultRequestHeaders.Authorization.Parameter);
         }
+
+
+        [Test]
+        public async Task TestBasicDownloadFileWithCancellationToken()
+        {
+            var batchId = Guid.NewGuid().ToString();
+            string destDownldLocation = Path.Combine(Environment.CurrentDirectory, "AFilename.txt");
```
Comment (crdandoukho) [inline Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs:136]: Please change all the new tests to use a MemoryStream for the destinationStream

File: Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs [hunk 8]
```diff
@@ -126,5 +127,116 @@ public async Task TestDownloadFileSetsAuthorizationHeader()
...
+        [Test]
+        public async Task TestForDownloadFilesWhenFileSizeIsGreaterThanMaxDownlodBytes()
+        {
+            var batchId = Guid.NewGuid().ToString();
+            string destDownldLocation = Path.Combine(Environment.CurrentDirectory, "AFilename.txt");
+            nextResponseStatusCode = HttpStatusCode.PartialContent;
+            var destStream = new FileStream(destDownldLocation, FileMode.OpenOrCreate, FileAccess.Write, FileShare.ReadWrite);
+            nextResponse = new MemoryStream(10585760);
+
+            var result = await fileShareApiClient.DownloadFileAsync(batchId, "AFilename.txt", destStream, 10585760, CancellationToken.None);
```
Comment (crdandoukho) [inline Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs:217]: When debugging this test it shows the range for the first chunk is 0 to 10485760 (10485761 bytes), which exceeds the `maxDownloadBytes` setting of 10485760. Could you please fix `DownloadFileAsync` to correctly honour the `maxDownloadBytes` setting.

I would also like to see this test changed to verify the correct ranges are being used for each download request

**Review by crdandoukho — CHANGES_REQUESTED** [review]
Thanks for your changes Rashi, though more changes are needed please.

Please see my comments

File: FileShareClient/FileShareApiClient.cs (hunk 6 above)
Comment (Rashi14748) [reply FileShareClient/FileShareApiClient.cs:156]: update same, to be handled by caller. Thanks

File: FileShareClient/FileShareApiClient.cs (hunk 5 above)
Comment (Rashi14748) [reply FileShareClient/FileShareApiClient.cs:115]: agree, updated as Stream. Thanks

File: Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs (hunk 7 above)
Comment (Rashi14748) [reply Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs:136]: changes done. Thanks

File: Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs (hunk 8 above)
Comment (Rashi14748) [reply Tests/UnitTests/FileShareClientTests/DownloadFilesTests.cs]: changes done for implementation. Thanks

File: FileShareClient/FileShareApiClient.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rashi14748) [reply FileShareClient/FileShareApiClient.cs]: changes done. Thanks

**Review by crdandoukho — APPROVED** [review]

**Review by barrie-cooper — APPROVED** [review]

**Pull request merged**